### PR TITLE
Add guide for configuring messaging in high availability for Wildfly

### PR DESCRIPTION
**Pull request by mnovak1**
Issue https://github.com/wildfly/wildfly.org/issues/623

Link to generated guide (located in my space)::   
https://mnovak1.github.io/guides/messaging-high-availability

### CODE REVIEW

File: guides/messaging-high-availability.adoc [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+[source, bash ,options="nowrap"]
+----
+mkdir messaging-journal-a
+mkdir messaging-journal-b
+----
+
+== Prepare WildFly Servers
+
+Now there will be configured two WildFly servers with two primary-secondary pairs. ActiveMQ Artemis configured
```
Comment (ehsavoie) [inline guides/messaging-high-availability.adoc:50]: Now let's configure two WildFly servers with two primary-secondary pairs of embedded ActiveMQ Artemis servers.

File: guides/messaging-high-availability.adoc [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+as primary on the first WildFly server and its secondary on the second WildFly server will use `messaging-journal-a`
```
Comment (ehsavoie) [inline guides/messaging-high-availability.adoc:51]: and as secondary on the second WildFly server

**Review by ehsavoie — COMMENTED** [review]
I have a question : you didn't define the connection factory or pcf to the cluster which should use the connectors to each broker

File: guides/messaging-high-availability.adoc [hunk 3]
```diff
@@ -0,0 +1,228 @@
+= Configure WildFly With Messaging (ActiveMQ Artemis) Cluster And High Availability
+:summary: Learn how to configure WildFly with messaging (ActiveMQ Artemis) cluster and high availability
+:includedir: _includes
+include::{includedir}/_attributes.adoc[]
+// you can override any attributes eg to lengthen the
+// time to complete the guide
+:prerequisites-time: 20
+
+In this guide you will learn how to configure two WildFly servers with messaging (ActiveMQ Artemis broker) in high availability topology with shared journal.
```
[the remark this answers is not in the document]
Comment (mnovak1) [reply guides/messaging-high-availability.adoc]: Fixed.

File: guides/messaging-high-availability.adoc [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+Apache ActiveMQ Artemis broker is integrated in WildFly and allows Artemis brokers to be linked together in primary-secondary pairs.
+It is active-passive high availability design such that one (the primary node) handles all the workload under normal operating conditions,
+while the other (the secondary node) remains on standby, ready to take over if the active node fails. As usually it’s not desirable to keep
+one Wildfly server idle on standby mode, we will configure one additional ActiveMQ Artemis server on each WidFly server as a secondary for the other.
```
[the remark this answers is not in the document]
Comment (mnovak1) [reply guides/messaging-high-availability.adoc]: I've tried to rephrase it to better describe the topology.

File: guides/messaging-high-availability.adoc (hunk 1 above)
Comment (mnovak1) [reply guides/messaging-high-availability.adoc]: Fixed.

File: guides/messaging-high-availability.adoc (hunk 2 above)
Comment (mnovak1) [reply guides/messaging-high-availability.adoc]: Fixed, thanks.

**Review by mnovak1 — COMMENTED** [review]
@ehsavoie thanks for the review. Fixed and repushed.

**Comment by mnovak1** [comment]
> I have a question : you didn't define the connection factory or pcf to the cluster which should use the connectors to each broker

@ehsavoie Actually it's not needed as once the connection factory or pcf creates connection to any broker, it receives topology of the whole cluster (even in case of in-vm connector in pcf). So it will know where the backup is and load balance all new connections to active brokers in cluster.

File: guides/messaging-high-availability.adoc [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+
+* Copy WildFly into two directories `wildfly-1` and `wildfly-2`
+* Start `wildfly-1` in `full-ha` profile in admin only mode and connect through CLI
+
+[source, bash ,options="nowrap"]
+----
+./wildfly-1/bin/standalone.sh -c standalone-full-ha.xml --admin-only
+----
+
+* In Another terminal connect to CLI:
```
Comment (ehsavoie) [inline guides/messaging-high-availability.adoc:68]: connect to the server with the jboss-cli

File: guides/messaging-high-availability.adoc [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+----
+
+* Start `wildfly-2` in `full-ha` profile in admin only mode and connect through CLI (set port offset to 1000 to avoid port conflicts)
+
+[source, bash ,options="nowrap"]
+----
+./wildfly-2/bin/standalone.sh -c standalone-full-ha.xml -Djboss.socket.binding.port-offset=1000 --admin-only
+----
+
+* In Another terminal connect to CLI:
```
Comment (ehsavoie) [inline guides/messaging-high-availability.adoc:132]: connect to the server with the jboss-cli

File: guides/messaging-high-availability.adoc [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+ready to take over if the primary broker fails. However, we will not start another WildFly server solely to support an inactive Artemis broker, instead we will configure an additional ActiveMQ Artemis broker
+on each WildFly instance to act as the secondary broker for the primary in the other WildFly. This setup results in two primary-secondary pairs, with each pair maintaining its own journal directory.
+
+image::artemis-collocated-topology.png[WildFly with ActiveMQ Artemis in collocated HA topology]
+
+If one WildFly server fails, the secondary ActiveMQ Artemis broker on the other WildFly server will activate and take over all duties.
+All HA enabled connections from the original primary broker will fail over to the secondary broker. The secondary broker will
+access all messages stored on the primary broker by loading them from the shared journal directory. Once the original WildFly server is restarted,
+the secondary broker will automatically switch back to standby mode, and all clients will fail back to the primary broker.
+Something like:
```
[the remark this answers is not in the document]
Comment (mnovak1) [reply guides/messaging-high-availability.adoc]: Removed, thanks :-)

File: guides/messaging-high-availability.adoc (hunk 5 above)
Comment (mnovak1) [reply guides/messaging-high-availability.adoc]: Fixed, thanks

File: guides/messaging-high-availability.adoc (hunk 6 above)
Comment (mnovak1) [reply guides/messaging-high-availability.adoc]: Fixed, thanks.

**Review by mnovak1 — COMMENTED** [review]
@ehsavoie thanks, fixed and repushed.

**Comment by bstansberry** [comment]
@ehsavoie Is this ok now?

Once approved please comment in https://wildfly.zulipchat.com/#narrow/stream/274477-Pull-requests/topic/wildfly.2Eorg.20PRs so we know to process it.

**Review by ehsavoie — APPROVED** [review]

File: _data/guides.yaml [hunk 8]
```diff
@@ -49,6 +49,12 @@ categories:
       - title: Using Credential Stores With Encrypted Expressions with WildFly
         url: /guides/security-credential-store-enc-exp
         description: How to set up encrypted expressions and use it to replace clear-text sensitive information.
+  - category: Messaging
+    cat-id: messaging
+    guides:
+      - title: Configuring Messaging in High Availability in Wildfly
```
Comment (bstansberry) [inline _data/guides.yaml:55]: I think these sound more natural:

"Configuring High Availability Messaging in WildFly"
"Configuring Messaging with High Availability in WildFly"

In both cases:

s/Wildfly/WildFly

File: _data/guides.yaml [hunk 9]
```diff
@@ -49,6 +49,12 @@ categories:
...
+        url: /guides/messaging-high-availability
+        description: Learn how to configure two WildFly servers with messaging (ActiveMQ Artemis broker) in high availability topology with shared journal.
```
Comment (bstansberry) [inline _data/guides.yaml:57]: s/in high/in a high/g

File: assets/img/news/artemis-collocated-topology.png
Comment (bstansberry) [inline assets/img/news/artemis-collocated-topology.png:1]: In the top two graphics

s/Wildfly/WildFly

File: guides/messaging-high-availability.adoc [hunk 10]
```diff
@@ -0,0 +1,227 @@
+= Configure WildFly With Messaging (ActiveMQ Artemis) Cluster And High Availability
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:1]: s/With/with a/g
s/And/and

File: guides/messaging-high-availability.adoc [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+:summary: Learn how to configure WildFly with messaging (ActiveMQ Artemis) cluster and high availability
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:2]: s/with/with a/g

File: guides/messaging-high-availability.adoc [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+:includedir: _includes
+include::{includedir}/_attributes.adoc[]
+// you can override any attributes eg to lengthen the
+// time to complete the guide
+:prerequisites-time: 20
+
+In this guide you will learn how to configure two WildFly servers with messaging (integrated ActiveMQ Artemis broker)
+in high availability topology with https://activemq.apache.org/components/artemis/documentation/latest/ha.html#shared-store[shared journal].
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:10]: s/in high/in a high/g
s/with shared journal/with a shared journal/g

File: guides/messaging-high-availability.adoc [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+image::artemis-collocated-topology.png[WildFly with ActiveMQ Artemis in collocated HA topology]
+
+If one WildFly server fails, the secondary ActiveMQ Artemis broker on the other WildFly server will activate and take over all duties.
+All HA enabled connections from the original primary broker will fail over to the secondary broker. The secondary broker will
+access all messages stored on the primary broker by loading them from the shared journal directory. Once the original WildFly server is restarted,
+the secondary broker will automatically switch back to standby mode, and all clients will fail back to the primary broker.
+
+[NOTE]
+--
+For simplicity, we will use just one machine to run WildFly instances and local directories for shared journals. In real world scenario those should be two separate machines with shared journal mount from shared file system (like NFS4).
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:26]: s/and local directories/and use local directories/g

The first 'use' verb is too far from the local directories part for it to be obvious that the local directories part is a 2nd object to the verb.

Comment (bstansberry) [reply guides/messaging-high-availability.adoc:26]: s/In real/In a real/g
s/with shared/with a shared/g
s/mount from/mounted from a/g

File: guides/messaging-high-availability.adoc [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+[source, bash ,options="nowrap"]
+----
+mkdir messaging-journal-a
+mkdir messaging-journal-b
+----
+
+== Prepare WildFly Servers
+
+Now let's configure two WildFly servers with two primary-secondary pairs of embedded ActiveMQ Artemis brokers. ActiveMQ Artemis configured
+as primary on the first WildFly server and as secondary on the second WildFly server will use `messaging-journal-a`
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:50]: s/will use/will use the/g
s/Similarly/Similarly, the/g
s/on second/on the second/g
s/and secondary/and the secondary/g

File: guides/messaging-high-availability.adoc [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+directory for their shared journal. Similarly primary broker on second WildFly server and secondary broker on
+the first WildFly server will use `messaging-journal-b` directory for their shared journal.
+
+[NOTE]
+--
+All ActiveMQ Artemis brokers must be in cluster so they know about each other and can provide cluster topology to connected clients. In case that any WildFly instance crashes, clients know the location of the secondary broker and are able to fail over.
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:56]: s/in cluster/in a cluster/g
s/In case that/In case/g

Comment (bstansberry) [reply guides/messaging-high-availability.adoc:56]: When I read this I feel like there's info I'm missing; i.e. they need to be in a cluster but it doesn't say how that is achieved, so then I feel anxiety.  Perhaps add something like "The procedures in this guide ensure the brokers are properly clustered."

File: guides/messaging-high-availability.adoc [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+--
+
+* Copy WildFly into two directories `wildfly-1` and `wildfly-2`
+* Start `wildfly-1` in `full-ha` profile in admin only mode and connect through CLI
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:60]: s/in `full-ha`/in a `full-ha`/g
s/admin only/admin-only/g

Drop 'and connect through CLI'. You have a separate step that does that.

File: guides/messaging-high-availability.adoc [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+[source, bash ,options="nowrap"]
+----
+./wildfly-1/bin/jboss-cli.sh -c
+----
+
+* Run following CLI commands on `wildfly-1`:
+
+[NOTE]
+--
+In the following CLI script change `<password>` and `<path>` values based on your environment.
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:78]: s/change/change the/g

File: guides/messaging-high-availability.adoc (hunk 16 above)
Comment (bstansberry) [reply guides/messaging-high-availability.adoc:60]: also, end the line with a ':'

File: guides/messaging-high-availability.adoc [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+
+# Set location of journal for pair A - change path to messaging-journal-a directory based on your environment
+/subsystem=messaging-activemq/server=default/path=bindings-directory:write-attribute(name=path,value=<path>/messaging-journal-a/bindings)
+/subsystem=messaging-activemq/server=default/path=journal-directory:write-attribute(name=path,value=<path>/messaging-journal-a/journal)
+/subsystem=messaging-activemq/server=default/path=large-messages-directory:write-attribute(name=path,value=<path>/messaging-journal-a/largemessages)
+/subsystem=messaging-activemq/server=default/path=paging-directory:write-attribute(name=path,value=<path>/messaging-journal-a/paging)
+
+# Set "default" Artemis broker as primary
+/subsystem=messaging-activemq/server=default/ha-policy=shared-store-primary:add(failover-on-server-shutdown=true)
+:reload
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:94]: s/:reload/reload/g

The string ':reload' should never appear in documentation. :) Users should always use the high-level command.

File: guides/messaging-high-availability.adoc [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+/subsystem=messaging-activemq/server=secondary/path=bindings-directory:write-attribute(name=path,value=<path>/messaging-journal-b/bindings)
+/subsystem=messaging-activemq/server=secondary/path=journal-directory:write-attribute(name=path,value=<path>/messaging-journal-b/journal)
+/subsystem=messaging-activemq/server=secondary/path=large-messages-directory:write-attribute(name=path,value=<path>/messaging-journal-b/largemessages)
+/subsystem=messaging-activemq/server=secondary/path=paging-directory:write-attribute(name=path,value=<path>/messaging-journal-b/paging)
+
+# Shutdown the server
+shutdown
+----
+
+* Start `wildfly-2` in `full-ha` profile in admin only mode and connect through CLI (set port offset to 1000 to avoid port conflicts)
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:124]: s/in full-ha/in a full-ha/g
s/admin only/admin-only/g

Drop 'and connect through CLI'. You have a separate step that does that.

end the line with a ':'

File: guides/messaging-high-availability.adoc [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+----
+
+* In a different terminal, connect to the server using the JBoss CLI:
+
+[source, bash ,options="nowrap"]
+----
+./wildfly-1/bin/jboss-cli.sh -c
+----
+
+* Run following CLI commands on `wildfly-1`:
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:74]: s/Run/Run the/g

File: guides/messaging-high-availability.adoc [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+----
+
+* In a different terminal, connect to the server using the JBoss CLI:
+
+[source, bash ,options="nowrap"]
+----
+./wildfly-2/bin/jboss-cli.sh -c --controller=127.0.0.1:10990
+----
+
+* Run following CLI commands on `wildfly-2`:
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:138]: s/Run/Run the/g

File: guides/messaging-high-availability.adoc [hunk 22]
```diff
@@ -0,0 +1,227 @@
...
+
+[NOTE]
+--
+In the following CLI script change `<password>` and `<path>` values based on your environment.
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:142]: s/change/change the/g

File: guides/messaging-high-availability.adoc [hunk 23]
```diff
@@ -0,0 +1,227 @@
...
+
+# Set location of journal for pair B - change path to messaging-journal-b directory based on your environment
+/subsystem=messaging-activemq/server=default/path=bindings-directory:write-attribute(name=path,value=<path>/messaging-journal-b/bindings)
+/subsystem=messaging-activemq/server=default/path=journal-directory:write-attribute(name=path,value=<path>/messaging-journal-b/journal)
+/subsystem=messaging-activemq/server=default/path=large-messages-directory:write-attribute(name=path,value=<path>/messaging-journal-b/largemessages)
+/subsystem=messaging-activemq/server=default/path=paging-directory:write-attribute(name=path,value=<path>/messaging-journal-b/paging)
+
+# Set "default" Artemis broker as primary
+/subsystem=messaging-activemq/server=default/ha-policy=shared-store-primary:add(failover-on-server-shutdown=true)
+:reload
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:158]: s/:reload/reload/g

File: guides/messaging-high-availability.adoc [hunk 24]
```diff
@@ -0,0 +1,227 @@
...
+/subsystem=messaging-activemq/server=secondary/path=large-messages-directory:write-attribute(name=path,value=<path>/messaging-journal-a/largemessages)
+/subsystem=messaging-activemq/server=secondary/path=paging-directory:write-attribute(name=path,value=<path>/messaging-journal-a/paging)
+
+# Shutdown the server
+shutdown
+----
+
+== Test High Availability
+
+We’ll test HA by crashing first WildFly server and checking that secondary ActiveMQ Artemis broker on the second WildFly server is active.
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:190]: s/crashing/crashing the/g
s/that secondary/that the secondary/g

File: guides/messaging-high-availability.adoc [hunk 25]
```diff
@@ -0,0 +1,227 @@
...
+
+* Start both WildFly servers, each in a separate terminal.
+
+[source, bash ,options="nowrap"]
+----
+./wildfly-1/bin/standalone.sh -c standalone-full-ha.xml
+./wildfly-2/bin/standalone.sh -c standalone-full-ha.xml -Djboss.socket.binding.port-offset=1000
+----
+
+* Shutdown or crash `wildfly-1` and check that secondary broker started in `wildfly-2` in the server log:
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:200]: s/that secondary/that the secondary/g

Comment (bstansberry) [reply guides/messaging-high-availability.adoc:200]: "check in the server log that the secondary broker started in `wildfly-2`:

File: guides/messaging-high-availability.adoc [hunk 26]
```diff
@@ -0,0 +1,227 @@
...
+
+[source, bash ,options="nowrap"]
+----
+16:25:44,921 INFO  [org.apache.activemq.artemis.core.server] (AMQ229000: Activation for server ActiveMQServerImpl::name=secondary) AMQ221010: Backup Server is now active
+----
+
+* Start `wildfly-1` server and verify that fail-back happened and secondary broker on `wildfly-2` went to standby mode:
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:207]: s/and secondary/and the secondary/g

File: guides/messaging-high-availability.adoc [hunk 27]
```diff
@@ -0,0 +1,227 @@
...
+[source, bash ,options="nowrap"]
+----
+16:26:53,043 INFO  [org.apache.activemq.artemis.core.server] (Thread-4 (ActiveMQ-scheduled-threads)) AMQ221008: primary server wants to restart, restarting server in backup
+...
+16:26:55,488 INFO  [org.apache.activemq.artemis.core.server] (Thread-0 (ActiveMQ-server-org.apache.activemq.artemis.core.server.impl.ActiveMQServerImpl$6@2845f281)) AMQ221031: backup announced
+----
+
+== What's next?
+
+ActiveMQ Artemis in WildFly allows also to use replicated journal where each primary-secondary pair replicates
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:218]: s/allows also/also allows you/g
s/replicated/a replicated/g

File: guides/messaging-high-availability.adoc [hunk 28]
```diff
@@ -0,0 +1,227 @@
...
+data over network. This has its pros and cons as it does not require to mount shared journal directories on
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:219]: s/over network/over the network/g

Comment (bstansberry) [reply guides/messaging-high-availability.adoc:219]: s/require to/require you to/g

File: guides/messaging-high-availability.adoc [hunk 29]
```diff
@@ -0,0 +1,227 @@
...
+each machine but usually suffers by lower performance caused by needed network round trip times between paired
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:220]: s/each machine/each machine,/g
s/by/from

File: guides/messaging-high-availability.adoc [hunk 30]
```diff
@@ -0,0 +1,227 @@
...
+brokers. More information can be found in https://activemq.apache.org/components/artemis/documentation/[ActiveMQ Artemis documentation].
```
Comment (bstansberry) [inline guides/messaging-high-availability.adoc:221]: s/found in/found in the/g

**Review by bstansberry — CHANGES_REQUESTED** [review]

File: guides/messaging-high-availability.adoc (hunk 12 above)
Comment (bstansberry) [reply guides/messaging-high-availability.adoc:10]: also, use...

[shared journal, window=_blank]

here and in other links, at least all those that appear in the main content.

That opens the link in a new tab, leaving the existing tab on the guide page.

File: assets/img/news/artemis-collocated-topology.png
Comment (mnovak1) [reply assets/img/news/artemis-collocated-topology.png:1]: Thanks, fixed.

**Review by mnovak1 — COMMENTED** [review]
@bstansberry Thanks a lot for the review! I've fixed all the issues and repushed. (also https://mnovak1.github.io/guides/messaging-high-availability is updated)

**Review by bstansberry — APPROVED** [review]

**Comment by bstansberry** [comment]
Thanks @mnovak1!

**Comment by mnovak1** [comment]
> Thanks @mnovak1!

@bstansberry @ehsavoie Thank you for your time and effort in reviewing my pull request. I really appreciate it!
